src/prefect/_internal/compatibility/experimental.py

```diff
@@ -0,0 +1,106 @@
+"""
+Utilities for experimental features.
+
+Experimental features have a group, feature name, and optional help string.
+
+When an experimental feature is used, a warning will be displayed. Warnings may be
+disabled per feature group with the setting `PREFECT_EXPERIMENTAL_WARN_<GROUP>`. 
+Warnings may also be disabled globally with the setting `PREFECT_EXPERIMENTAL_WARN`.
+
+Some experimental features require opt-in to enable any usage. These require the setting
+`PREFECT_EXPERIMENTAL_ENABLE_<GROUP>` to be set or an error will be thrown on use.
+"""
+import functools
+import warnings
+from typing import Callable, TypeVar
+
+from prefect.settings import PREFECT_EXPERIMENTAL_WARN, SETTING_VARIABLES, Setting
+
+T = TypeVar("T", bound=Callable)
+
+EXPERIMENTAL_WARNING = (
+    "{feature} is experimental and the interface or behavior may change without warning. "
+    "We recommend pinning to a specific version to avoid breaking changes. "
+    "{help} "
+    "To disable warnings for this group, disable PREFECT_EXPERIMENTAL_WARN_{group}. "
+)
+
+EXPERIMENTAL_ERROR = (
+    "{feature} is experimental and requires opt-in for usage. "
+    "{help} "
+    "To use this feature, enable PREFECT_EXPERIMENTAL_ENABLE_{group}. "
+)
+
+
+class ExperimentalWarning(Warning):
+    """
+    A warning for use of expirimental code.
+
+    These can be globally disabled by the PREFECT_EXPIRIMENTAL_WARN setting.
+    """
+
+
+class ExperimentalError(Exception):
+    """
+    An exception for use of experimental code.
+    """
+
+
+def _opt_in_setting_for_group(group: str) -> Setting[bool]:
+    group_opt_in_setting_name = f"PREFECT_EXPERIMENTAL_ENABLE_{group.upper()}"
+    group_opt_in = SETTING_VARIABLES.get(group_opt_in_setting_name)
+    if group_opt_in is None:
+        raise ValueError(
+            f"A opt-in setting for experimental feature {group!r} does not exist yet. "
+            f"{group_opt_in_setting_name!r} must be created before the group can be "
+            "used."
+        )
+    return group_opt_in
+
+
+def _warn_setting_for_group(group: str) -> Setting[bool]:
+    group_warn_setting_name = f"PREFECT_EXPERIMENTAL_WARN_{group.upper()}"
+    group_warn = SETTING_VARIABLES.get(group_warn_setting_name)
+    if group_warn is None:
+        raise ValueError(
+            f"A warn setting for experimental feature {group!r} does not exist yet. "
+            f"{group_warn_setting_name!r} must be created before the group can be used."
+        )
+    return group_warn
+
+
+def experimental(
+    group: str, feature: str, help: str = "", stacklevel: int = 2, opt_in: bool = False
+) -> Callable[[T], T]:
```

Comment:
Decorator typing is already kinda tricky, and decorator _factory_ typing is even more whacky.  I think you need to go beyond just `Callable` and use `ParamSpec` in order for the function types to carry through.  We have some examples in Nebula where we're doing that: https://github.com/PrefectHQ/nebula/blob/main/src/prefect_cloud/utilities/tracing.py#L111-L117

Reply:
I _think_ most of this is possible in 3.8 but you'll have to monkey about with it

Reply:
Really? I think that's only needed if you're moving things around

<img width="225" alt="Screen Shot 2022-12-09 at 12 19 43 PM" src="https://user-images.githubusercontent.com/2586601/206767157-d55bff21-d3ce-4e4e-b26d-a9a0421a517f.png">

Here, we're saying this function returns a callable which given a type T (which is also a callable in this case) returns type T so if you pass it a function with some signature it returns that whole thing back "unchanged".

Reply:
Whoa I think you just helped me simplify some things 🤯

Reply:
Pyright likes it; let's hope mypy does too!